mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxFactory.kt

```diff
@@ -1,10 +1,139 @@
 package com.airbnb.mvrx
 
+import android.os.Bundle
+import android.os.Parcelable
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.ViewModelProvider
 import androidx.annotation.RestrictTo
+import java.io.Serializable
+import java.lang.IllegalStateException
+import kotlin.reflect.full.primaryConstructor
 
 @RestrictTo(RestrictTo.Scope.LIBRARY)
-class MvRxFactory<V>(private val factory: (Class<*>) -> V) : ViewModelProvider.Factory {
-    @Suppress("UNCHECKED_CAST") override fun <T : ViewModel?> create(modelClass: Class<T>) = factory(modelClass) as T
-}
\ No newline at end of file
+
+class MvRxFactory<VM : BaseMvRxViewModel<S>, S : MvRxState>(
+    private val viewModelClass: Class<out VM>,
+    private val stateClass: Class<out S>,
+    private val viewModelContext: ViewModelContext,
+    private val key: String,
+    private val forExistingViewModel: Boolean = false,
+    private val initialStateFactory: MvRxStateFactory<VM, S> = RealMvRxStateFactory()
+) : ViewModelProvider.Factory {
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
+        val restoredState = viewModelContext.savedStateRegistry.consumeRestoredStateForKey(key)
+
+        val (viewModelContext, stateRestorer) =
+            restoredState
+                ?.toRestoredState(viewModelContext)
+                ?: viewModelContext to { state: S -> state }
+
+        val viewModel = createViewModel(
+            viewModelClass,
+            stateClass,
+            viewModelContext,
+            key,
+            stateRestorer,
+            forExistingViewModel = restoredState == null && forExistingViewModel,
+            initialStateFactory = initialStateFactory
+        )
+
+        viewModelContext.savedStateRegistry.registerSavedStateProvider(key) {
+            viewModel.getSavedStateBundle(viewModelContext)
+        }
+
+        return viewModel as T
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+private fun <VM : BaseMvRxViewModel<S>, S : MvRxState> createViewModel(
+    viewModelClass: Class<out VM>,
+    stateClass: Class<out S>,
+    viewModelContext: ViewModelContext,
+    key: String,
+    stateRestorer: (S) -> S,
+    forExistingViewModel: Boolean,
+    initialStateFactory: MvRxStateFactory<VM, S>
+): VM {
+    if (forExistingViewModel) {
```

Comment:
I feel like it is a little odd to have a parameter which if true throws an exception -- could the caller perform this check? This keeps the function more true to it's name `createViewModel`.